feat(doc): add guest configuration to bao-classic config documentation

Diogo21Costa opened the pull request:
## PR Description

In this pull request, I'm adding content to the "Guest Configuration" section of the config documentation for the Bao hypervisor.

1. Guest Image Configuration
2. VM Configuration (Virtual Platform)
3. CPU Affinity (Mapping of Virtual to Physical CPUs)
4. Cache Coloring Setup

The focus here is to provide a comprehensive description for each field, offering detailed insights into their functionalities.

Do you have any suggestions or feedback concerning the content provided in the "Guest Configuration" section?

Diogo21Costa committed:
feat(doc): add config description

Signed-off-by: Diogo21Costa <diogoandreveigacosta@gmail.com>

josecm commented:
I think  @DavidMCerdeira should also participate in these reviews...

josecm commented on source/bao_hyp/config.rst:
```diff
@@ -68,36 +68,280 @@ relative to the config source file.
+where:
+
+- **base**  [mandatory] - corresponds to the virtual base address of the IPC memory region;
+- **size** [mandatory] - corresponds to the size of the IPC memory region;
```
Also, for MPU systems, base is ignored (the region address is the shared memory object address)

Diogo21Costa replied:
thanks, addressed in 404de5a52494eed38516059edb32e4e7031331a3.

josecm commented on source/bao_hyp/config.rst:
```diff
@@ -68,36 +68,280 @@ relative to the config source file.
+- **cpu_affinity** [optional] - corresponds to a bitmap signaling the preferred physical CPUs
+  assigned to the VM. If this value is mutually exclusive for all the VMs, the physical CPUs
+  assigned to each VM follow the bitmap. Otherwise (in case of bit overlap or lack of affinity
+  definition), the CPU assignment is defined by the hypervisor;
```
the cpu affinity will be as best as possible be followed, but it is not guaranteed.

Diogo21Costa replied:
Thanks, added this information!

josecm reviewed (changes requested):
I think we need a few iterations on this one. And possibly discuss an alternative structure where we go into details and have text explaining the features and functionalities, while here we only go field by field as you did.

Diogo21Costa commented on source/bao_hyp/config.rst:
```diff
@@ -68,36 +68,280 @@ relative to the config source file.
+    :name: guest-config-fig
+
+
+The configuration of different guests is done by populating a struct called *vmconfig*, as follows:
```
Thanks, fixed in 923bbca5719fda58025bfc192b7461efa1d1a1c4.

Diogo21Costa commented on source/bao_hyp/config.rst:
```diff
@@ -68,36 +68,280 @@ relative to the config source file.
+.. code-block:: c
+
+    struct vm_config {
+        struct image;
```
Thanks, fixed in fadc7a2bb90adc5f6d6fd4eebad36337f2748c39.

Diogo21Costa commented on source/bao_hyp/config.rst:
```diff
@@ -68,36 +68,280 @@ relative to the config source file.
+- **load_addr** [mandatory] - corresponds to the ``image`` load address in the hypervisor address
+  space. This value can be defined using the macro VM_IMAGE_OFFSET(img_name);
+- **size** [mandatory] - corresponds to the image size. This value can be defined using the macro
+  VM_IMAGE_SIZE(img_name);
```
Thanks, fixed in dde0f11fddf139eee0aa33223d587b912c60aecd.

Diogo21Costa commented on source/bao_hyp/config.rst:
```diff
@@ -68,36 +68,280 @@ relative to the config source file.
+            size_t size;
+            bool separately_loaded;
+            bool inplace;
+    };
```
Thanks, fixed in fadc7a2bb90adc5f6d6fd4eebad36337f2748c39.

Diogo21Costa commented on source/bao_hyp/config.rst:
```diff
@@ -68,36 +68,280 @@ relative to the config source file.
 Virtual Machine Configuration
 *****************************
 
+The VM configuration enables users to define the characteristics of each the virtualized platform.
```
Thanks, fixed in 27f6cb74ebdae864bb7417b908a9de0f35853215.

Diogo21Costa commented on source/bao_hyp/config.rst:
```diff
@@ -68,36 +68,280 @@ relative to the config source file.
+- **Inter-Process Comunication (IPC)** - see details in `Inter-Process Communication (IPC)`_;
+- **Devices** - see details in `Devices`_;
+- **Memory Management** - see details in `Memory Management`_;
+- **Architectural-Specific Configurations** - see details in `Architectural-Specific
```
Fully agree, refactored this in 3fec42951b45f15f0dad7da83f26a5fc75f07994.

Diogo21Costa commented on source/bao_hyp/config.rst:
```diff
@@ -68,36 +68,280 @@ relative to the config source file.
 Guests Configuration
```
Thanks, fixed in c4e19ed96a14cf7fe0af614d016e1dec60a51fb3.

Diogo21Costa commented on source/bao_hyp/config.rst:
```diff
@@ -68,36 +68,280 @@ relative to the config source file.
+environment, including hardware resources, devices, and storage options. One of the most important
+features of the configuration file is the ability to define VMs, or guests (VM 1, VM 2, ... VM N).
+This allows to create and manage multiple VMs, each with its own operating system and hardware
+resources.
```
Thanks, fixed in 93be85f1070b26f5c6757f2f814d989021e213b4.

DavidMCerdeira replied:
Contradaction between "by assigning them to one or more VMs." and "In Bao, resources are exclusively assigned to each VM"

Diogo21Costa replied:
Agree! Fixed in 0af22168fee64bff57dd3bd95ee208fe42ea6bba.

Diogo21Costa commented on source/bao_hyp/config.rst:
```diff
@@ -68,36 +68,280 @@ relative to the config source file.
+    :name: guest-config-fig
+
+
+The configuration of different guests is done by populating a struct called *vmconfig*, as follows:
```
Thanks, fixed in 923bbca5719fda58025bfc192b7461efa1d1a1c4.

DavidMCerdeira replied:
I see the latest update still uses the term guest

Diogo21Costa replied:
Thanks! Fixed in c7a6d758b17018485011afcf2ae6c825513b7f38

Diogo21Costa commented on source/bao_hyp/config.rst:
```diff
@@ -68,36 +68,280 @@ relative to the config source file.
+  Machine Configuration`_);
+- **cpu_affinity** [optional] - corresponds to the selection of physical CPUs assigned to the
+  virtual platform (see details in `CPU Affinity`_);
+- **colors** [optional] - enables the cache coloring feature (see details in `Coloring`_).
```
Thanks! Fixed in 7ab7bf436542d6fd6626211577837d6ecc5452d1.

Diogo21Costa commented on source/bao_hyp/config.rst:
```diff
@@ -68,36 +68,280 @@ relative to the config source file.
+The guest ``image`` comprises a structure that describes the image configuration running on the
+guest side. It encompasses the following options:
+
+- **image** [mandatory] - definition of the ``image`` to run on a given VM. The ``image``
+  corresponds to the following structure:
```
While this information may seem redundant, providing context about the image within the vm_config struct can be valuable for readers. Additionally, incorporating a pointer or link to review any necessary information could assist individuals engaging in a non-linear read. For instance, if someone specifically seeks the VM configuration section and bypasses the image definition, having this contextualization might be helpful. Should we remove this and keep it simpler (users still have the table of contents on the Read The Docs page)?

Diogo21Costa commented on source/bao_hyp/config.rst:
```diff
@@ -68,36 +68,280 @@ relative to the config source file.
+encompasses the definition of:
+
+- **Number of virtual CPUs** - see details in `Number of vCPUs`_;
+- **Virtual memory regions** - see details in `Memory Mapping`_;`
```
Thanks, fixed in 50d0eec13ce3fa0bf1c39a368e078985a0135d19.

Diogo21Costa commented on source/bao_hyp/config.rst:
```diff
@@ -68,36 +68,280 @@ relative to the config source file.
+.. _Number of vCPUs:
+
+- **cpu_num** [mandatory] - defines the number of CPUs assigned to the VM;
+
```
Good point, fixed in a0259bc0ac7777e5bd1c04970f47d6ef1a52a9f2.

Diogo21Costa commented on source/bao_hyp/config.rst:
```diff
@@ -68,36 +68,280 @@ relative to the config source file.
+- **cpu_num** [mandatory] - defines the number of CPUs assigned to the VM;
+
 
 2. Memory Mapping
```
Thanks, fixed in b3489c1663dcc54035af65c2b1aba0165b852434.

Diogo21Costa commented on source/bao_hyp/config.rst:
```diff
@@ -68,36 +68,280 @@ relative to the config source file.
+.. _Memory Mapping:
+
+- **region_num** [mandatory] - defines the number of memory regions mapped to the VM. This
+  structure contemplates the following parameters:
```
Thanks, fixed in f8011d97198f8b3d9f6a4683c582bd316dd9e5c1.

Diogo21Costa commented on source/bao_hyp/config.rst:
```diff
@@ -68,36 +68,280 @@ relative to the config source file.
 
 2. Memory Mapping
 #################
+.. _Memory Mapping:
```
Agree, addressed in 1b77d8a1ea233e229e576cb103ea2fc5e3cbd45a.

DavidMCerdeira replied:
The document still misses a definition of guest address vs physical address

Diogo21Costa replied:
Agree! Added in 2aa913b901de0b0d8d55c073a2a10e24ace91ba8.

Diogo21Costa commented on source/bao_hyp/config.rst:
```diff
@@ -68,36 +68,280 @@ relative to the config source file.
+  memory region. By default, ``place_phys`` equals to false;
+- **phys** [mandatory if ``place_phys`` is true] - it corresponds to the physical address where the
+  memory region should be mapped;
 
```
Good points, thank you! Added in c5923099c8b4ab784b4bad4c305a0668063606e3 and in 9e61aa497068aead1e3aa1c1f492a34c39f0c275.

Diogo21Costa commented on source/bao_hyp/img/llc-colors.svg:
Fully agree, updated in e501eb4f6d92ed2ecac512ce6b2356c3b49caccf.

DavidMCerdeira replied:
Might be worth it to also color the main memory

Diogo21Costa replied:
Updated in 96b43946f86ca07200eeef477e02d495f7896aed.

DavidMCerdeira replied:
Looks good
Only thing left is to fix the colors they should be different for VM1 and VM2

Diogo21Costa replied:
Thank you @DavidMCerdeira ! Fixed the image in commit https://github.com/bao-project/bao-docs/pull/78/commits/820da47b5498b9f3b4658293fdf9713252bc8d0c.

Diogo21Costa commented on source/bao_hyp/config.rst:
```diff
@@ -68,36 +68,280 @@ relative to the config source file.
+
+- **vgic_dscrp** [mandatory] - corresponds to the configuration of the Generic Interrupt Controller
+  (GIC);
+- **vgic_dscrp** [optional] - corresponds to the configuration of the System Memory Management Unit
```
Thanks, fixed in 431be7fe3361d849639a1aa0ccacdae07fc7eb24.

Diogo21Costa commented on source/bao_hyp/config.rst:
```diff
@@ -68,36 +68,280 @@ relative to the config source file.
+        } gic;
+
+        // Configuration of the System Memory Management Unit (SMMU)
+        struct {
```
Thanks, fixed in e9e1c43697985f49f1c3c7b0400b42b10e280401.

Diogo21Costa commented on source/bao_hyp/config.rst:
```diff
@@ -68,36 +68,280 @@ relative to the config source file.
+To ease the process of configuring the image running on each VM, the configuration of Bao allows
+the use of two different macros:
+
+1. **VM_IMAGE_BUILTIN** - This macro simplifies image configuration by requiring only the
+   ``img_name`` and the image ``base_addr``. This macro specifies both the base address and image
+   size.
+
+2. **VM_IMAGE_LOADED** - This macro requires additional configurations. It requires the definition
+   of image ``base_addr``, the image ``load_addr``, and the image ``size``.
```
Thanks, addressed in dd49512744034522cd1abd99b0207ef2c5b862bf.

Diogo21Costa commented on source/bao_hyp/config.rst:
```diff
@@ -68,36 +68,280 @@ relative to the config source file.
+    :name: vm-image-fig
+
+To ease the process of configuring the image running on each VM, the configuration of Bao allows
+the use of two different macros:
```
Thanks, fixed in 987878a6aa4e515a104be67c7cb9152d3c0fe7d5.

DavidMCerdeira replied:
Add a warning or a note box to highlight this better

Diogo21Costa replied:
Thanks! Fixed in f7e00438422e29f58fd532f4ed3eefc69fc15ce2.

Diogo21Costa commented on source/bao_hyp/img/vm-image.svg:
Thanks, removed the image as it was not conveying the correct idea.

Diogo21Costa commented on source/bao_hyp/config.rst:
```diff
@@ -68,36 +68,280 @@ relative to the config source file.
+
+- **base_addr** [mandatory] - corresponds to the ``image`` load address in the VM's address space;
+- **load_addr** [mandatory] - corresponds to the ``image`` load address in the hypervisor address
+  space. This value can be defined using the macro VM_IMAGE_OFFSET(img_name);
```
Thanks, fixed in 4b057cd11f289d87d6c70c3ef205118ce1610361.

Diogo21Costa commented on source/bao_hyp/config.rst:
```diff
@@ -68,36 +68,280 @@ relative to the config source file.
+where:
+
+- **pa** [mandatory] - corresponds to the base physical address of the device;
+- **va** [mandatory] - corresponds to the base virtual address of the device;
+- **size** [mandatory] - corresponds to the size of the device memory region;
```
Thanks, addressed in ae5a5bdb1904aeea420d94356fc72ceb0cc801f3.

Diogo21Costa commented on source/bao_hyp/config.rst:
```diff
@@ -68,36 +68,280 @@ relative to the config source file.
+- **phys** [mandatory if ``place_phys`` is true] - it corresponds to the physical address where the
+  memory region should be mapped;
 
 3. Inter-Process Communication (IPC)
 ####################################
+.. _Inter-Process Communication (IPC):
```
Agree, included new content in 854fe14d5edc08d3e958ff0fa3f458baa11507d1.

Diogo21Costa commented on source/bao_hyp/config.rst:
```diff
@@ -68,36 +68,280 @@ relative to the config source file.
+
+where:
+
+- **base**  [mandatory] - corresponds to the virtual base address of the IPC memory region;
+- **size** [mandatory] - corresponds to the size of the IPC memory region;
```
Thanls, fixed in 404de5a52494eed38516059edb32e4e7031331a3.

Diogo21Costa commented on source/bao_hyp/config.rst:
```diff
@@ -68,36 +68,280 @@ relative to the config source file.
+  assigned to each VM follow the bitmap. Otherwise (in case of bit overlap or lack of affinity
+  definition), the CPU assignment is defined by the hypervisor;
+
 Coloring
 ********
```
Thanks, addressed in 2bbd78f5573c827b7df785a6c983c26983be633c.

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -65,39 +65,404 @@ describe the image running on the VM (see `Guest Image`_);
 --------------------
 
+Bao's configuration file allows you to partition the platforms' hardware resources, such as CPU
+cores, memory, or devices, by assigning them one or more VMs. It also allows you to configure the
```
```suggestion
cores, memory, or devices, by assigning them to one or more VMs. It also allows you to configure the
```

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -65,39 +65,404 @@ describe the image running on the VM (see `Guest Image`_);
+    :name: guest-config-fig
+
+
+Each entry in the ``vmlist`` mentioned earlier is a ``vm_config`` struct, defining the
+configuration for different guests by populating the following struct called *vmconfig*:
```
```suggestion
Each entry in the ``vmlist`` mentioned earlier is a ``vm_config`` struct, which defines the
configuration of each individual guest:
```

Diogo21Costa replied:
Thanks, fixed!

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -65,39 +65,404 @@ describe the image running on the VM (see `Guest Image`_);
+CPU affinity, color mapping, and platform details. For each VM, the following parameters must be
+specified:
+
+- **image** [mandatory] - a structure containing information about guest image loading (see details
+  in `Guest Image`_)
+- **entry** [mandatory] - defines the entry point address in VM's address space;
+- **platform description** [mandatory] - a description of the VM platform, defining its resource
+  assignments and requirements (see details in `Virtual Machine Configuration`_);
+- **cpu_affinity** [optional] - defines the affinity of the VM's vCPUs to the physical CPUs
+  assigned to the virtual platform (see details in `CPU Affinity`_);
+- **colors** [optional] - assignment of shared LLC cache colors (or partitions) to this VM (see
+  details in `Coloring`_).
```
Use the inline code format for each code related keyword (struct fields, macros, etc).

Diogo21Costa replied:
Thanks, fixed!

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -65,39 +65,404 @@ describe the image running on the VM (see `Guest Image`_);
+correct syntax and parameters. Attempting to manually configure image details without utilizing
+these macros may result in errors or unintended behavior.
+
+The integration of the appropriate macro tailored to your specific use case, encompassing essential
+parameters such as img_name, base_addr, load_addr, and size. This practice not only fosters
+consistency but also enhances reliability in your VM setup. For instance:
```
The writing here is really intricate and not so clear.

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -65,39 +65,404 @@ describe the image running on the VM (see `Guest Image`_);
+true, the guest image is not embedded in the hypervisor image; instead, it is loaded independently.
+For more details, refer to the figure below.
+
+.. figure:: img/guest-image.svg
+    :align: center
+    :name: vm-image-fig
```
To make the scale to the width of the window or the image seems to small.
```suggestion
.. figure:: img/guest-image.svg
    :align: center
    :width: 100%
    :name: vm-image-fig
```

Diogo21Costa replied:
Thanks, fixed!

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -65,39 +65,404 @@ describe the image running on the VM (see `Guest Image`_);
 
+The VM configuration enables users to define the characteristics of each virtualized platform. It
+encompasses critical details that define the VM's run-time environment, performance, and overall
+capabilities. The virtual machine configuration is performed by populating the structure `struct
+vm_platform`, outlined below:
```
```suggestion
capabilities. The virtual machine configuration is performed by populating the structure ``struct
vm_platform``, outlined below:
```

Diogo21Costa replied:
Thanks, fixed!

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -65,39 +65,404 @@ describe the image running on the VM (see `Guest Image`_);
+    }
+
+By customizing this configuration, users can tailor the virtual platform to suit specific workload
+requirements and application needs for their virtual machines. The configuration encompasses the
```
```suggestion
requirements and application needs for their virtual machines. The configuration includes the
```

Diogo21Costa replied:
Thanks, fixed!

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -65,39 +65,404 @@ describe the image running on the VM (see `Guest Image`_);
 ##################
+.. _Number of vCPUs:
+
+- **cpu_num** [mandatory] - defines the number of CPUs assigned to the VM;
```
Use inline format for the field name. Normalize on the overall document.

Diogo21Costa replied:
Thanks, fixed!

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -65,39 +65,404 @@ describe the image running on the VM (see `Guest Image`_);
+physical memory for these regions. However, there is an option to define the physical memory region
+where the specific region will be mapped.
+
+For each VM, users can define multiple memory regions. To facilitate this, users first define the
+total number of memory regions via the `region_num` parameter:
+
+- **region_num** [mandatory] - defines the number of memory regions in the VM, specifically, the
+  number of `vm_mem_region` entries in the `vm_platform`'s `regions` list.
+
+Then, each memory region is described by populating the `struct vm_mem_region`:
```
```suggestion
For each VM, users can define multiple memory regions. To facilitate this, users first define the
total number of memory regions via the ``region_num`` parameter:

- ``region_num`` [mandatory] - defines the number of memory regions in the VM, specifically, the
  number of ``vm_mem_region`` entries in the ``vm_platform``'s ``regions`` list.

Then, each memory region is described by populating the ``struct vm_mem_region``:
```

Diogo21Costa replied:
Thanks, fixed!

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -65,39 +65,404 @@ describe the image running on the VM (see `Guest Image`_);
+  address. If ``place_phys`` equals to true, it allows to specify the physical address of the
+  memory region. By default, ``place_phys`` equals to false;
+
+.. note::
+  In MPU systems, place_phys and phys are ignored.
```
Remove. Already below.

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -65,39 +65,404 @@ describe the image running on the VM (see `Guest Image`_);
+- **phys** [mandatory if ``place_phys`` is true] - it corresponds to the physical address where the
+  memory region should be mapped;
+
+.. note::
+  For enhanced performance, especially in MMU-based targets, it's recommended to align `base` and
+  `size` to the architecture's huge pages (e.g., 2 MiB for Arm and RISC-V). Similarly, if
+  `place_phys` is enabled, aligning `phys` to the architecture's huge pages can also improve
+  performance.
+
+.. note::
+  In MPU systems, place_phys and phys are ignored.
```
Merge the two notes into one. Make a list inside the note for both points.

Diogo21Costa replied:
Thanks, fixed!

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -65,39 +65,404 @@ describe the image running on the VM (see `Guest Image`_);
+fostering enhanced functionality and system efficiency.
+
+.. figure:: img/ipc.svg
+    :align: center
+    :name: ipc-fig
```
```suggestion
    :align: center
    :width: 100%
    :name: ipc-fig
```

Diogo21Costa replied:
Thanks, fixed!

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -65,39 +65,404 @@ describe the image running on the VM (see `Guest Image`_);
+    :align: center
+    :name: ipc-fig
+
+Bao provides support for IPC, allowing VMs to establish communication channels. The IPC
```
To add space below the image and text.
```suggestion
|
```

Diogo21Costa replied:
Thanks, fixed!

danielRep replied:
You need to use `|` to produce the empty blank lines. Fix it in other images.

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -65,39 +65,404 @@ describe the image running on the VM (see `Guest Image`_);
+involves selecting the physical core where the guest should run.
+
+.. figure:: img/cpu-affinity.svg
+    :align: center
+    :name: cpu-affinity-fig
```
```suggestion
    :align: center
    :width: 100%
    :name: cpu-affinity-fig
```

Diogo21Costa replied:
Thanks, fixed!

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -65,39 +65,404 @@ describe the image running on the VM (see `Guest Image`_);
+    :align: center
+    :name: cpu-affinity-fig
+
+This functionality is achieved through the following configuration parameter:
```
Add an empty line between image and text.

Diogo21Costa replied:
Added, thanks!

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -65,39 +65,404 @@ describe the image running on the VM (see `Guest Image`_);
+  picture depicts a hypothetical setup with a 50/50 coloring scheme;
+
+.. figure:: img/llc-colors.svg
+    :align: center
+    :name: llc-colors-fig
```
```suggestion
    :align: center
    :width: 100%
    :name: llc-colors-fig
```

Diogo21Costa replied:
Thanks, fixed!

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -65,39 +65,404 @@ describe the image running on the VM (see `Guest Image`_);
+.. figure:: img/llc-colors.svg
+    :align: center
+    :name: llc-colors-fig
+
```
add empty line

Diogo21Costa replied:
Fixed!

danielRep replied:
See comment above.

danielRep reviewed (changes requested):
Major points to fix:
- Add width directive on each figure to 100% to fill the window size;
- Use the inline code format correctly and in all code-related text;
- Clear some sentences more intricate.

Diogo21Costa commented on source/bao_hyp/config.rst:
```diff
@@ -68,36 +68,280 @@ relative to the config source file.
+ cortex-r) the hypervisor sets up the VM using an MPU by default. If the user wants this VM to use
+ the MMU they must set the config ``mmu`` parameter to true;
 
 6. Architectural-Specific Configurations
```
Removed old section 5 and split the old section 6 (currently in section 5) in the two architectures (arm and risc-v). However, I saw this information was already on the porting guide, so I mirrored the information here.

danielRep reviewed (changes requested):
Fix the space between text and figure directives, and it LGTM.

Diogo21Costa commented:
> Fix the space between text and figure directives, and it LGTM.

Sorry for the delay. I missed this comment. Fixed in https://github.com/bao-project/bao-docs/pull/78/commits/7bfac37f5444434e623341767dc631d076f2a6ce.

DavidMCerdeira commented on source/bao_hyp/config.rst:
```diff
@@ -65,39 +65,468 @@ describe the image running on the VM (see `Guest Image`_);
+.. code-block:: c
+
+    struct vm_config {
+        struct vm_image image;
```
This is slightly different now:

struct {
        vaddr_t base_addr;
        paddr_t load_addr;
        size_t size;
        bool separately_loaded;
        bool inplace;
    } image;

There's no struct type for vm_image

Diogo21Costa replied:
Thanks! Fixed in https://github.com/bao-project/bao-docs/pull/78/commits/8a6069e2433f9c4946b4b5b41c74bb117b1f01fa!

DavidMCerdeira commented on source/bao_hyp/config.rst:
```diff
@@ -65,39 +65,468 @@ describe the image running on the VM (see `Guest Image`_);
+
+- ``IMAGE_LOADED``: Highly recommended, especially for MPU systems, where manual allocation of
+  space for the guest image can be challenging if embedded in Bao's binary. Without utilizing
+  LOADED, Bao may need to copy the image, potentially resulting in space wastage.
```
```suggestion
  LOADED, Bao may need to copy the image, potentially resulting in wasted space.
```

Diogo21Costa replied:
Fixed in [commit2](https://github.com/bao-project/bao-docs/pull/78/commits/00a28ffd02c8edeac156fa2d0859626846edc820)!

DavidMCerdeira commented on source/bao_hyp/config.rst:
```diff
@@ -65,39 +65,468 @@ describe the image running on the VM (see `Guest Image`_);
+true, the guest image is not embedded in the hypervisor image; instead, it is loaded independently.
+For more details, refer to the figure below.
+
+.. figure:: img/guest-image.svg
+    :align: center
+    :width: 100%
+    :name: vm-image-fig
+
+|
```
Good job with the figure!

Diogo21Costa replied:
Thanks!

DavidMCerdeira commented on source/bao_hyp/config.rst:
```diff
@@ -65,39 +65,468 @@ describe the image running on the VM (see `Guest Image`_);
 #################
+.. _Memory Regions:
+
+Bao employs a two-stage translation mechanism to manage VM memory. Each VM memory region is
```
Not true for mpu though

DavidMCerdeira replied:
I'd remove the first paragraph

Diogo21Costa replied:
You're right, removed in https://github.com/bao-project/bao-docs/pull/78/commits/69735d900ae91470b9b688cd750092215ff595f4!

DavidMCerdeira commented on source/bao_hyp/config.rst:
```diff
@@ -65,39 +65,468 @@ describe the image running on the VM (see `Guest Image`_);
 
 3. Inter-Process Communication (IPC)
 ####################################
+.. _Inter-Process Communication (IPC):
```
Should be inter-partition communication

Diogo21Costa replied:
Fully agree, fixed in [commit4](https://github.com/bao-project/bao-docs/pull/78/commits/e564d6e6f90ab61d7657c4118f1b55f674690370).

DavidMCerdeira commented on source/bao_hyp/config.rst:
```diff
@@ -65,39 +65,468 @@ describe the image running on the VM (see `Guest Image`_);
+
+Inter-Process Communication (IPC) enables seamless communication between distinct processes in a
+computing system. This mechanism facilitates data exchange, synchronization, and coordination,
+fostering enhanced functionality and system efficiency.
```
I'm okay with leaving as is, but I don't like that it reads as straight chatGPT output

Diogo21Costa replied:
I agree with the point, so I rewrote this part to make it simpler (https://github.com/bao-project/bao-docs/pull/78/commits/696ae378ab4327d6c353941cfb76d0c8d5c05d10). What are your thoughts on this approach?

DavidMCerdeira replied:
It's ok for me

DavidMCerdeira commented on source/bao_hyp/config.rst:
```diff
@@ -65,39 +65,468 @@ describe the image running on the VM (see `Guest Image`_);
+
+Bao provides support for IPC, allowing VMs to establish communication channels. The IPC
+configuration involves defining the number of IPCs using the ``ipc_num`` field within the
+``vm_config`` struct. The specifics of each IPC are then outlined through the ``ipcs`` structure,
```
```suggestion
``vm_platform`` struct. The specifics of each IPC are then outlined through the ``ipcs`` structure,
```

Diogo21Costa replied:
Thanks! Fixed in [commit6](https://github.com/bao-project/bao-docs/pull/78/commits/a83bb96520b8955efd336b16238d573fb23e92fb).

DavidMCerdeira commented on source/bao_hyp/config.rst:
```diff
@@ -65,39 +65,468 @@ describe the image running on the VM (see `Guest Image`_);
+
+where:
+
+- ``base``  [mandatory] - corresponds to the virtual base address of the IPC memory region;
```
This is a general comment:
To make thinks clearer we should use terms like `guest physical address`, or `guest address`
Here I suggest replacing `virtual address` with `guest address`, and be consistent across the document
thoughts @danielRep @josecm ?

danielRep replied:
I tend to agree with this.

Diogo21Costa replied:
Agree, fixed in https://github.com/bao-project/bao-docs/pull/78/commits/831e0663c86cd0e53ff5411f8c67aa12de02777e.

DavidMCerdeira commented on source/bao_hyp/config.rst:
```diff
@@ -65,39 +65,468 @@ describe the image running on the VM (see `Guest Image`_);
+.. code-block:: c
+
+    struct vm_dev_region {
+        paddr_t pa;
+        vaddr_t va;
```
Currently device configuration is awkward for MPU configuration as there is no va

I suggest to state this clearly

We should work on this at some point

Diogo21Costa replied:
Thanks! Tried to address this point by updating the note in https://github.com/bao-project/bao-docs/pull/78/commits/3558ec91451e0f20fb210ad2178de2abc561258f, but I'm not sure if it fully addresses this point.

DavidMCerdeira reviewed (changes requested):
Looks good to me, apart from minor issues

DavidMCerdeira commented on source/bao_hyp/config.rst:
```diff
@@ -65,39 +65,468 @@ describe the image running on the VM (see `Guest Image`_);
+specific memory region. Cache coloring exclusively operates in virtual memory systems, i.e.,
+systems featuring Memory Management Units (MMUs) for address translation.
+
 Shared Memory Configuration
 ---------------------------
```
Remove this here

Diogo21Costa replied:
Fixed in https://github.com/bao-project/bao-docs/pull/78/commits/d2efdcf6c8ba853a19100408e1b1620feef7fa85!

DavidMCerdeira commented on source/bao_hyp/config.rst:
```diff
@@ -363,8 +363,9 @@ where:
-  For MMU systems, this typically aligns to 4K, while for MPU systems, it aligns to 64 bytes.
+  It is mandatory for ``base`` and ``size`` to align with the smallest page size of the
+  architecture. For MMU systems, this typically aligns to 4K, while for MPU systems, it aligns to
+  64 bytes. Please note that for MPU systems there is no virtual address (VA).
```
Currently, the VA address must be set anyway for MPU systems

Diogo21Costa replied:
Thanks for clarifying, updated in https://github.com/bao-project/bao-docs/pull/78/commits/a139c992247e67ff35c787efe7afd279a15a7ff7!

DavidMCerdeira commented on source/bao_hyp/config.rst:
```diff
@@ -529,6 +529,46 @@ However, this mechanism may not take effect if the physical mapping feature is e
+---------------------------
+.. _Shared Memory Configuration:
+
+In Bao's configuration multiple (``shmem0``, ``shmem1``, ...) or no shared memory can be set. If
+at least one shared memory is defined, an  ID is assigned to which the shared memory is
+associated. Later, this ID can be associated with an IPC in the multi-guest configuration.
```
```suggestion
In Bao's configuration multiple (``shmem0``, ``shmem1``, ...) or no shared memory can be set. An  ID is assigned to each shared memory object. Later, this ID can be linked to an IPC in the multi-guest configuration.
```

Diogo21Costa replied:
Thanks! Fixed  in commit https://github.com/bao-project/bao-docs/pull/78/commits/33652e5f3d7a0867ce9cda4be518f5496cf1078b.

DavidMCerdeira commented on source/bao_hyp/config.rst:
```diff
@@ -529,6 +529,46 @@ However, this mechanism may not take effect if the physical mapping feature is e
+
+|
+
+The configuration of shared memories is done by populating a struct called
+``shmem``, as follows:
```
```suggestion
The configuration of shared memoy is done by populating a struct called
``shmem``. The struct definition is presented bellow:
```

Diogo21Costa replied:
Thanks! Fixed in commit https://github.com/bao-project/bao-docs/pull/78/commits/b4d14d8ed94fcbddc9aa54c21ecef8d0eb53111b.

DavidMCerdeira commented on source/bao_hyp/config.rst:
```diff
@@ -529,6 +529,46 @@ However, this mechanism may not take effect if the physical mapping feature is e
+
+- **size** [mandatory] - defines the total size of the shared memory. The size of the shared \
+  memory must always be multiples of 4Kb (0x1000);
+- **place_phys** [optional] - shared memory is mapped into the virtual memory and it's \
+  important to note that the virtual address (VA) might not necessarily be the same as the \
+  physical address (PA). When ``place_phys`` is set to true, the virtual address corresponds to \
+  the physical address. By default, ``place_phy`` equals to false;
+- **base / phys** [optional] - defines the physical address (base address) of the shared memory.
```
```suggestion
- **place_phys** [optional] - shared memory is mapped into the virtual memory and it's \
  important to note that the guest address might not necessarily be the same as the \
  physical address (PA). When ``place_phys`` is set to true, the guest address corresponds to \
  the physical address. By default, ``place_phy`` equals to false;
- **base / phys** [optional] - defines the physical address (base address) of the shared memory.
```

Diogo21Costa replied:
Thanks! Fixed in commit https://github.com/bao-project/bao-docs/pull/78/commits/f6b8cded246f578fef61aaf115ffd77e7988f732.

DavidMCerdeira commented on source/bao_hyp/config.rst:
```diff
@@ -529,6 +529,46 @@ However, this mechanism may not take effect if the physical mapping feature is e
+
+- **size** [mandatory] - defines the total size of the shared memory. The size of the shared \
+  memory must always be multiples of 4Kb (0x1000);
+- **place_phys** [optional] - shared memory is mapped into the virtual memory and it's \
+  important to note that the virtual address (VA) might not necessarily be the same as the \
+  physical address (PA). When ``place_phys`` is set to true, the virtual address corresponds to \
+  the physical address. By default, ``place_phy`` equals to false;
+- **base / phys** [optional] - defines the physical address (base address) of the shared memory.
```
This is not clear. Please clarify that when place_phys is false base address is the guest address, otherwise phys should be used to set the shared memory physical address

Diogo21Costa replied:
Fixed in commit https://github.com/bao-project/bao-docs/pull/78/commits/07452d7d6faf6f7bd6e5205d03c04c05c84938f6 (updated both the ``place_phys``and the ``base / phys`` descriptions.

DavidMCerdeira commented on source/bao_hyp/config.rst:
```diff
@@ -65,41 +65,512 @@ describe the image running on the VM (see `Guest Image`_);
 ---------------------------
+.. _Shared Memory Configuration:
+
+In Bao's configuration multiple (``shmem0``, ``shmem1``, ...) or no shared memory can be set. An ID
```
```suggestion
In Bao's configuration, you can set multiple shared memory regions (e.g., shmem0, shmem1) or none at all.
```

Diogo21Costa replied:
Thanks, fixed in commit https://github.com/bao-project/bao-docs/pull/78/commits/3f2a6f174933062acb335c27d1a70adca084d5bc!

DavidMCerdeira reviewed (commented):
Looks good to me.
Only thing missing is re-writing the history which should be done after addressing all pending conversations

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -65,41 +65,512 @@ describe the image running on the VM (see `Guest Image`_);
+The guest ``image`` comprises a structure that describes the image configuration running on the
+guest side. It encompasses the following options:
+
+- **image** [mandatory] - definition of the ``image`` to run on a given VM. The ``image``
```
```suggestion
- ``image`` [mandatory] - definition of the ``image`` to run on a given VM. The ``image``
```

Diogo21Costa replied:
Thanks! Fixed in https://github.com/bao-project/bao-docs/pull/78/commits/12a303c36d4dc3d6310e65291c125de5729336a5!

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -65,41 +65,512 @@ describe the image running on the VM (see `Guest Image`_);
+requirements and application needs for their virtual machines. The configuration includes the
+definition of:
+
+- ``Number of CPUs`` - see details in `Number of vCPUs`_;
+- ``Memory regions`` - see details in `Memory Regions`_;`
+- ``Inter-Partition Comunication (IPC)`` - see details in `Inter-Partition Communication (IPC)`_;
+- ``Devices`` - see details in `Devices`_;
+- ``Architectural-Specific Configurations`` - see details in `Architectural-Specific
```
```suggestion
- **Number of CPUs** - see details in `Number of vCPUs`_;
- **Memory regions** - see details in `Memory Regions`_;
- **Inter-Partition Comunication (IPC)** - see details in `Inter-Partition Communication (IPC)`_;
- **Devices** - see details in `Devices`_;
- **Architectural-Specific Configurations** - see details in `Architectural-Specific
```

Diogo21Costa replied:
Thanks! Fixed in https://github.com/bao-project/bao-docs/pull/78/commits/9ffd91e472802b0f7d18b6ea36be3344f52e904f!

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -65,41 +65,512 @@ describe the image running on the VM (see `Guest Image`_);
+
+- ``dev_num`` [mandatory] - corresponds to the number of devices assigned to the VM;
+- ``devs`` [mandatory if *dev_num* > 0] - corresponds to the specification of the VM's devices and
+  is configured through the following structure:
```
```suggestion
- ``devs`` [mandatory if ``dev_num`` > 0] - corresponds to the specification of the VM's devices and
```

Diogo21Costa replied:
Thanks! Fixed in https://github.com/bao-project/bao-docs/pull/78/commits/b7787266d1b74ec1e430a6d98ad1429d0b1e16aa!

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -65,41 +65,512 @@ describe the image running on the VM (see `Guest Image`_);
+
+This struct contains the following parameters:
+
+- **size** [mandatory] - defines the total size of the shared memory. The size of the shared \
+  memory must always be multiples of 4Kb (0x1000);
+- **place_phys** [optional] - This flag determines the mapping of shared memory into virtual \
+  memory. When ``place_phys`` is set to ``false`` (default value), the base address is a guest \
+  (virtual) address. When set to ``true``, the guest address directly corresponds to the physical \
+  address (PA) of the shared memory.
+- **base / phys** [optional] - When ``place_phys`` is set to ``false``, the ``base`` parameter \
```
```suggestion
- ``size`` [mandatory] - defines the total size of the shared memory. The size of the shared \
  memory must always be multiples of 4Kb (0x1000);
- ``place_phys`` [optional] - This flag determines the mapping of shared memory into virtual \
  memory. When ``place_phys`` is set to ``false`` (default value), the base address is a guest \
  (virtual) address. When set to ``true``, the guest address directly corresponds to the physical \
  address (PA) of the shared memory.
- ``base`` / ``phys`` [optional] - When ``place_phys`` is set to ``false``, the ``base`` parameter \
```

Diogo21Costa replied:
Thanks! Fixed in https://github.com/bao-project/bao-docs/pull/78/commits/9744a25a97ea9dbb532bb1011ae33889d826cd17!

danielRep reviewed (commented):
Apart from minor issues easily solved, it looks very good to me! Great work Diogo.

DavidMCerdeira commented on source/bao_hyp/config.rst:
```diff
@@ -65,41 +65,512 @@ describe the image running on the VM (see `Guest Image`_);
+- ``size`` [mandatory] - corresponds to the size of the device memory region;
 
-6. Architectural-Specific Configurations
+.. note::
```
Add a second note that we shouldn't add the gic to these devices.
A scenario would be where we want to give a large portion of the MMIO, and that MMIO range contains the GIC. In this case we need to split the MMIO range in two creating a hole for the gic.

Diogo21Costa replied:
Thank you for the review @DavidMCerdeira! Addressed this point in [this commit](https://github.com/bao-project/bao-docs/pull/78/commits/0c360b5e516e0cb6233d261d7a7877bcb5ec1676)

DavidMCerdeira commented on source/bao_hyp/config.rst:
```diff
@@ -483,6 +483,16 @@ where:
 
+.. note::
+
+   When mapping MMIO regions for guests, the memory regions associated with the GIC (Generic
+   Interrupt Controller) must be excluded. Mapping these regions can lead to conflicts or incorrect
+   behavior, as they are typically managed by Bao through trap-and-emulate mechanisms.
```
When mapping MMIO regions for guests, the memory regions associated with the interrupt controller (only interrupt controllers are target of MMIO trap and emulate in Bao) must be excluded. Mapping these regions can lead to conflicts or incorrect
   behavior, as they are typically managed by Bao through trap-and-emulate mechanisms.

Diogo21Costa replied:
Thank you @DavidMCerdeira, updated in commit https://github.com/bao-project/bao-docs/pull/78/commits/7cc5013e494154afea5728bfca456de3de34749b.

Diogo21Costa commented on source/bao_hyp/config.rst:
```diff
@@ -68,36 +68,280 @@ relative to the config source file.
 ***********
+.. _Guest Image:
+
+The guest ``image`` comprises a structure that describes the image configuration running on the
+guest side. It encompasses the following options:
```
Agree! Addressed in https://github.com/bao-project/bao-docs/pull/78/commits/e46aa17c7c8adcfc8fe1c2ae55fc9153981868e9 & https://github.com/bao-project/bao-docs/pull/78/commits/521c3399c5f327593e34c6b01440d11d99a51876.

Diogo21Costa commented on source/bao_hyp/config.rst:
```diff
@@ -68,36 +68,280 @@ relative to the config source file.
+  the VM. By default, ``interrupt_num`` equals to 0;
+- **interrupts** [mandatory if *interrupt_num*>0] - defines a list of interrupt IDs generated by
+  the device - ``(irqid_t[]) {irq_1, ..., irq_n};``
+- **id** [optional] - corresponds to the bus master id for iommu effects:
```
Thanks! Addressed in https://github.com/bao-project/bao-docs/pull/78/commits/373c222729f9776e64484936fe5da0e1491527ca

DavidMCerdeira commented on source/bao_hyp/img/cpu-affinity.svg:
0bxxx is not portable behavior across compilers
Please use is hexadecimal instead, i.e., 0x1, 0x2, 0xC,

Diogo21Costa replied:
Thanks! Fixed in bec0caea4b1438c44b561f562c4bc4c8a3a05d51.

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -49,57 +49,587 @@ for each list, it is necessary to specify the list size using the parameters ``s
+
+Bao's configuration file allows you to partition the platform’s hardware resources, including CPU
+cores, memory regions, devices, and communication channels, across multiple VMs according to your
+needs. In Bao, CPU cores, memory regions, and interrupt are always assigned exclusively to a
```
```suggestion
needs. In Bao, CPU cores, memory regions, and interrupts are always assigned exclusively to a
```

danielRep replied:
You missed this one @Diogo21Costa

Diogo21Costa replied:
You're right, sorry! Fixed in https://github.com/bao-project/bao-docs/pull/78/commits/9b3a4f7a55dec41f30407e088573b4d190286a27.

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -49,57 +49,587 @@ for each list, it is necessary to specify the list size using the parameters ``s
+  assignments and requirements (see details in `Virtual Machine Configuration`_);
+- ``cpu_affinity`` [optional] - defines the affinity of the VM's vCPUs to the physical CPUs
+  assigned to the virtual platform. The affinity will be followed as best as possible, but it is
+  not guaranteed (see details in `CPU Affinity`_).;
```
```suggestion
  not guaranteed (see details in `CPU Affinity`_);
```

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -49,57 +49,587 @@ for each list, it is necessary to specify the list size using the parameters ``s
+- ``load_addr`` [mandatory] - corresponds to the ``image`` load ``physical address``. This value
+  can be defined using the macro ``VM_IMAGE_OFFSET(img_name)``;
+- ``size`` [mandatory] - corresponds to the image size. For builtin images declared using
+  `VM_IMAGE`, this value can be defined using the macro ``VM_IMAGE_SIZE(img_name)``;
```
```suggestion
  ``VM_IMAGE``, this value can be defined using the macro ``VM_IMAGE_SIZE(img_name)``;
```

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -49,57 +49,587 @@ for each list, it is necessary to specify the list size using the parameters ``s
+- ``size`` [mandatory] - corresponds to the image size. For builtin images declared using
+  `VM_IMAGE`, this value can be defined using the macro ``VM_IMAGE_SIZE(img_name)``;
+- ``separately_loaded`` [optional] - informs the hypervisor if the VM image is to be loaded
+  separately by a bootloader; By default, separately_loaded is set as false;
```
```suggestion
  separately by a bootloader. By default, ``separately_loaded`` is set as ``false``;
```

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -49,57 +49,587 @@ for each list, it is necessary to specify the list size using the parameters ``s
+- ``separately_loaded`` [optional] - informs the hypervisor if the VM image is to be loaded
+  separately by a bootloader; By default, separately_loaded is set as false;
+- ``inplace`` [optional]- use the image inplace and don’t copy the image. By default, inplace is
+  set as false;
```
```suggestion
- ``inplace`` [optional]- use the image inplace and don’t copy the image. By default, ``inplace`` is
  set as ``false``;
```

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -49,57 +49,587 @@ for each list, it is necessary to specify the list size using the parameters ``s
+``VM_IMAGE_OFFSET``. During run-time, the hypervisor adjusts this value to be subsequently
+interpreted as a physical address. This adjustment involves adding the address at which the
+hypervisor itself was loaded. However, if the ``separately_loaded`` parameter is configured as
+true, the VM image is not embedded in the hypervisor image; instead, it is loaded independently.
```
```suggestion
``true``, the VM image is not embedded in the hypervisor image; instead, it is loaded independently.
```

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -49,57 +49,587 @@ for each list, it is necessary to specify the list size using the parameters ``s
+  details in `Coloring`_).
+
+VM Image
+***********
```
```suggestion
VM Image
********
```

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -49,57 +49,587 @@ for each list, it is necessary to specify the list size using the parameters ``s
+the `VM Image`_ section.
+
+VM Configuration
+--------------------
```
```suggestion
VM Configuration
----------------
```

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -49,57 +49,587 @@ for each list, it is necessary to specify the list size using the parameters ``s
 #################
+.. _Memory Regions:
+
+In Bao’s configuration, you will encounter both ``guest addresses`` and ``physical addresses``
```
```suggestion
In Bao’s configuration, you will encounter both **guest addresses** and **physical addresses**
```

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -49,57 +49,587 @@ for each list, it is necessary to specify the list size using the parameters ``s
+  important to note that the virtual address (VA) might not necessarily be the same as the physical
+  address (PA). When "place_phys" is set to true, the VM address corresponds to the physical
+  address. If ``place_phys`` equals to true, it allows to specify the physical address of the
+  memory region. By default, ``place_phys`` equals to false;
```
```suggestion
  memory region. By default, ``place_phys`` equals to ``false``;
```

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -49,57 +49,587 @@ for each list, it is necessary to specify the list size using the parameters ``s
+  address. If ``place_phys`` equals to true, it allows to specify the physical address of the
+  memory region. By default, ``place_phys`` equals to false;
+
+- ``phys`` [mandatory if ``place_phys`` is true] - it corresponds to the physical address where the
```
```suggestion
- ``phys`` [mandatory if ``place_phys`` is ``true``] - it corresponds to the physical address where the
```

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -49,57 +49,587 @@ for each list, it is necessary to specify the list size using the parameters ``s
+      and ``size`` to the architecture's huge pages (e.g., 2MiB for Arm and RISC-V). Similarly, if
+      ``place_phys`` is enabled, aligning ``phys`` to the architecture's huge pages can also
+      improve performance.
+    - In MPU systems, place_phys and phys are ignored.
```
```suggestion
    - In MPU systems, ``place_phys`` and ``phys`` are ignored.
```

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -49,57 +49,587 @@ for each list, it is necessary to specify the list size using the parameters ``s
+
+- ``shmem_id`` [mandatory] - corresponds to the ID of the shared memory associated with the IPC;
+- ``interrupt_num`` [mandatory] - defines the number of interrupts assigned to the IPC;
+- ``interrupts`` [mandatory if *interrupt_num* > 0] - defines a list of interrupt IDs assigned to
```
```suggestion
- ``interrupts`` [mandatory if ``interrupt_num`` > 0] - defines a list of interrupt IDs assigned to
```

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -49,57 +49,587 @@ for each list, it is necessary to specify the list size using the parameters ``s
+where:
+
+- ``pa`` [mandatory] - corresponds to the base ``physical address`` of the device;
+- ``va`` [mandatory] - corresponds to the base VM ``physical address`` of the device;
```
Physical?

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -49,57 +49,587 @@ for each list, it is necessary to specify the list size using the parameters ``s
+
+- ``interrupt_num`` [optional] - corresponds to the number of interrupts generated by the device to
+  the VM. By default, ``interrupt_num`` equals to 0;
+- ``interrupts`` [mandatory if *interrupt_num*>0] - defines a list of interrupt IDs generated by
```
```suggestion
- ``interrupts`` [mandatory if ``interrupt_num``>0] - defines a list of interrupt IDs generated by
```

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -49,57 +49,587 @@ for each list, it is necessary to specify the list size using the parameters ``s
+      Redistributor;
+    - ``gic.maintenance_id`` [mandatory] - The interrupt ID for the GIC's maintenance interrupt;
+
+    For the SMMU `struct smmu_dscrp`:
```
```suggestion
    For the SMMU ``struct smmu_dscrp``:
```

danielRep commented on source/bao_hyp/config.rst:
```diff
@@ -49,57 +49,587 @@ for each list, it is necessary to specify the list size using the parameters ``s
+  (virtual) address. When set to ``true``, the VM address directly corresponds to the physical \
+  address (PA) of the shared memory.
+- ``base / phys`` [optional] - When ``place_phys`` is set to ``false``, the ``base`` parameter \
+  should be used to specify the VM virtual address of the shared memory. When place_phys is \
```
```suggestion
  should be used to specify the VM virtual address of the shared memory. When ``place_phys`` is \
```

danielRep reviewed (commented):
Small formating typos. LGTM. Great job!

Diogo21Costa commented:
> Small formating typos. LGTM. Great job!

Thank you @danielRep for the review! Addressed all comments in commit https://github.com/bao-project/bao-docs/pull/78/commits/d25e7ac1ddbbc2728c1285ea8be5bd2ef7e2b49b.
